### DISCUSSION TITLE
Wireless Microphone Architecture for Birdnet-Pi and Insetnet-Pi?

### DISCUSSION THREAD

**Discussion by Astrobirder** [mcguirepr89/BirdNET-Pi#574] [Ideas]
Hello,

At the suggestion of [@nilspupils](https://github.com/nilspupils), I'm splitting out this discussion from @DD4WH 's Microphone thread.

I've been considering a different architecture for my birdnet-pi installation from a microphone standpoint: placing the RPi indoors and connecting the microphones wirelessly. This would allow for a better environment for the RPi (don't have to deal with high heat in summer, especially for the RPi4, as well as a greater degree of electronic and audio isolation: electronically from a power supply noise standpoint and audio wise with respect to potentially having a fan cool the pi. It would also facilitate updating the pi and make maintenance easier.

Specifically I was thinking about having the microphone connected to an ESP32 or possibly arduino. As for transport, I'd use either WiFi (RTSP) or possibly BLE with LC3 if it turns out to be good enough from a quality standpoint. I know that there is a port of the LC3 for Arduino that was done by Phil Schatzmann (https://github.com/pschatzmann/arduino-liblc3). The ESP32/arduino could then be powered with solar/battery setups. I have experience with Bluetooth classic and the rather poor SBC codec, so I'm a bit leery of this approach, so I'd appreciate the thoughts of the community. Maybe Frank (@DD4WH) or Nils (@nilspupils) has played with these?

Another interesting idea would be to potentially use the auracast feature of LE Audio to allow for various birdnet pi's to listen to the same microphone(s) with different learning models.  For example adding an insect-focused model instead of the bird-focused model yet only having to deploy a single microphone setup.

Ideas? Suggestions?

**Comment by matthew73210** [comment]
> Hello,
> 
> At the suggestion of [@nilspupils](https://github.com/nilspupils), I'm splitting out this discussion from @DD4WH 's Microphone thread.
> 
> I've been considering a different architecture for my birdnet-pi installation from a microphone standpoint: placing the RPi indoors and connecting the microphones wirelessly. This would allow for a better environment for the RPi (don't have to deal with high heat in summer, especially for the RPi4, as well as a greater degree of electronic and audio isolation: electronically from a power supply noise standpoint and audio wise with respect to potentially having a fan cool the pi. It would also facilitate updating the pi and make maintenance easier.
> 
> Specifically I was thinking about having the microphone connected to an ESP32 or possibly arduino. As for transport, I'd use either WiFi (RTSP) or possibly BLE with LC3 if it turns out to be good enough from a quality standpoint. I know that there is a port of the LC3 for Arduino that was done by Phil Schatzmann (https://github.com/pschatzmann/arduino-liblc3). The ESP32/arduino could then be powered with solar/battery setups. I have experience with Bluetooth classic and the rather poor SBC codec, so I'm a bit leery of this approach, so I'd appreciate the thoughts of the community. Maybe Frank (@DD4WH) or Nils (@nilspupils) has played with these?
> 
> Another interesting idea would be to potentially use the auracast feature of LE Audio to allow for various birdnet pi's to listen to the same microphone(s) with different learning models. For example adding an insect-focused model instead of the bird-focused model yet only having to deploy a single microphone setup.
> 
> Ideas? Suggestions?

Hey @Astrobirder

At the moment, because of some pulseaudio issues with my orange pi. I use an old rpi3 to stream via rstp over the network, and connect via the rstp stream option in configs with and address like this "tcp://birdnetmicrophone:5002". Works pretty well, but sometimes if wifi goes down, birdnet will go haywire and stop just store all the wav files without actually analysing them.

**Reply by Astrobirder** [reply]
@matthew73210 Thanks for that!  It's exactly the type of info that I'm looking for here, so we can weigh the various wireless options.  I'm aware that we can currently change the birdnet-pi to "listen" to an rtsp stream (thanks for that @mcguirepr89 !), and that's certainly one possibility.

**Comment by nilspupils** [comment]
You probably have seen these already, if not they might be worth a look:
https://github.com/pschatzmann/Micro-RTSP-Audio
https://github.com/atomic14/esp32_wireless_microphone
https://github.com/atomic14/esp32-walkie-talkie
https://github.com/vernonet/ESP32_PRJ

**Comment by DD4WH** [comment]
Great idea!

The large disadvantage of a wireless mic is -of course- that you need a separate power supply in your garden/backyard etc.
But if you solve that, I could imagine that a ESP32 LyraT could do the job. Some thoughts on this:
* the LyraT has a dedicated codec with a built-in mic preamp. So you theoretically should be able to get decent audio quality from your mic and get it into the ESP32 in digital format
* then, the ESP32 LyraT can transmit the audio via Bluetooth or via rstp stream
* the ESP32 LyraT costs about 25€ at mouser, so its pretty cheap and you would not need any additional hardware

Regarding different models for different species groups: Well, there is only ONE model for the BirdNET-Pi: BirdNET-Lite and thats for birds.
I do not know of any other model compatible with BirdnET-Pi.
And it would be an extremely large effort to train models for other groups.
But we should not discuss that here, that would be a suggestion for Stefan Kahl who trains the models:
https://github.com/kahst/BirdNET-Analyzer

I would buy an ESP32 LyraT and experiment with it. I have some experience in microcontrollers, but have never used ESP32, so I expect a rather steep learning curve ;-)

**Reply by Astrobirder** [reply]
@DD4WH Hmm I've not played with the LyraT before, but I do have a couple of other ESP32 devices (WROOM32).  I'll have to look at those, especially since I'm about to place a Mouser order for the PUI 5024 microphones (Hope I'm picking the right one--4 choices: https://www.mouser.com/c/?q=pui%205024 .  I have the 665-AOM-5024L-HD-F-R version in my cart.

As for the power supply, I was thinking battery + solar as a completely wireless option, but I also have mains voltage available outside--so I'm just avoiding having the Pi outside or trying to run an audio or USB cable from inside to outside.

Hmm, lots of choices for kinds of LyraT devices as well.  Is there a specific one you were suggesting?  I guess I'll search for one with the dedicated codec & preamp.

**Reply by DD4WH** [reply]
Hmm, not sure ab out which AOM5024 to choose . . . I would pick the cheapest one.

I am not familiar with ESP32 LyraT, I just noticed from the datasheet that it has a codec onboard (I am pretty sure this is independent of the version).  https://espressif-docs.readthedocs-hosted.com/projects/esp-adf/en/latest/design-guide/dev-boards/board-esp32-lyrat-v4.3.html#other-versions-of-lyrat

The codec is an ES8388 with two mic inputs (including bias for the electret mic and a preamp). That is the main reason I thought it could be useable, because then you would not need any additional hardware.

**Reply by Astrobirder** [reply]
Not sure the LyraT is the way to go for my first tests.  It doesn't appear to support the LC3 codec required for BLE LE Audio.  I'm really thinking auracast BLE microphones might be the way to go, to facilitate switching microphones.  Speaking of that, I know that the BirdnetPi system can use either a USB microphone/soundcard or an RTSP stream.  If I want to connect a pi to some type of BLE receiver (as BLE 5.2 or 5.3 is required), is it easy to pipe in an I2S stream? @mcguirepr89? I've not done that kind of move. 

Maybe one of the NRF5340 boards might be better suited for that purpose, but I've not played with them at all.  Anyone else?

**Comment by kinteki** [comment]
The cheapest:
Using an old, but working android phone. Install _[IP Webcam](https://play.google.com/store/apps/details?id=com.pas.webcam&hl=hu&gl=US)_ . You can use its audio rtsp channel for distant microphone. Fix IP on android phone. It works around the house wifi( for example in the garden). Plus: you can use it as webcam too.

**Comment by nilspupils** [comment]
Hey guys have a look at this: 
https://github.com/umr-ds/BirdEdge-Mic
They use a wireless mic for their application.

**Comment by mmcc-xx** [comment]
Hi All - I just built a remote microphone using an example found here: https://github.com/pschatzmann/arduino-audio-tools
It streams via http rather than rtsp, but the ffmpeg command used to do stream recording doesn't seem to care.

I used this example: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/sandbox/streams-i2s-webserver_aac
and I used the same microphone he describes in the readme.

You'll need to install this library as well: https://github.com/pschatzmann/arduino-fdk-aac

**Reply by Astrobirder** [reply]
Thanks for sharing!  

Let us know if you connect it to your BirdNET-Pi and what kind of results you get!

**Reply by mmcc-xx** [reply]
It seems to work... ok. The mic isn't the best and I'm looking into how I might attach a better one. The streaming aspect works fine. 

Here's the spectrogram of a northern flicker it picked up. I tried uploading the audio, but mp3s are not permitted to be attached.

![image](https://user-images.githubusercontent.com/1315341/235367587-f1bf7784-065b-4a1a-bb28-c9d24c2aa953.png)

**Reply by kobius77** [reply]
there seem to be (few) boards like this - https://www.audiophonics.fr/en/devices-hifi-audio-adc/adc-analog-to-digital-converter-wm8782-i2s-24bit-192khz-p-13351.html - which _should_ work as well, shouldn't it? as i understand, I2S is like serial for audio.

this also might be of interest https://www.youtube.com/watch?v=a936wNgtcRA

anyways, i'm after the same thing, a low power networked microphone delivering a stream that a birdnetpi-instance can consume. just started doing so with a webcam, which works surprisingly well..

**Comment by mmcc-xx** [comment]
The device you linked to _should_ work. I have a board on its way to me from China that has a 3.5mm jack that I'm planning to connect a cheapo lav mic to, and maybe gin up a parabolic dish for it. I found some web page showing the use of a dished round plastic garbage can lid for that purpose.

https://www.amazon.com/dp/B09CDCLXQC?psc=1&ref=ppx_yo2ov_dt_b_product_details
https://www.amazon.com/dp/B09BN6KGMT?psc=1&ref=ppx_yo2ov_dt_b_product_details
https://www.videomaker.com/article/f20/17144-how-to-build-a-parabolic-mic-dish/

The library mentioned in the video you linked to is the one I'm using. It seems to work well. The author of the library suggested not using AAC, but rather using something that requires less horsepower to encode like SBC. His library has an SBC encoder but I haven't tried it out yet - waiting for the board.

The mic built into my camera works OK but the audio stream is a pretty low sampling rate. Higher sampling rate should mean better capture of high frequencies.

I mentioned above but I'll mention it again - rtsp does not seem to be required - a straight http stream works fine. There might be more lag between when the sound is received and when it was produced but that doesn't seem critical for this use case.

**Reply by kobius77** [reply]
> I mentioned above but I'll mention it again - rtsp does not seem to be required - a straight http stream works fine.

yes, which might make things easier, but not sure if rtsp still would be better suited (but haven't bothered checking for how to actually create such stream with an esp) - citing some random google result: "RTSP will not drop the connection, no matter what state it is in. HTTP, on the other hand, does not save the state."

**Comment by NateEaton** [comment]
Anyone here have updates to share on progress over the past 7-8 months? This is _exactly_ what I am interested in having given the extreme summer temperatures we had in North Texas last year.

**Comment by blogdemoi** [comment]
What is Insetnet-Pi?

Thank you.
